Controller to add/remove/rename user roles

litvinovg opened the pull request:
The issue: [4079](https://github.com/vivo-project/VIVO/issues/4079)

# What does this pull request do?
Created role controller to create/remove, rename roles.

# What's new?
Roles controller

# How should this be tested?
In admin panel try to create/remove/rename roles

# Interested parties
@VIVO-project/vivo-committers

# Reviewers' expertise

Candidates for reviewing this PR should have some of the following expertises:
1. Java
2. HTML, CSS, JavaScript
3. FreeMarker
4. SPARQL
5. Natural language knowledge
    1. English
    2. German
    3. Spanish
    4. French
    5. Portuguese
    6. Russian
    7. Serbian

# Reviewers' report template
## General comment
A reviewer should provide here comments and suggestions for requested changes if any.
## Testing
A reviewer should briefly describe here how it was tested
## Code reviewing
A reviewer should briefly describe here which part was code reviewed

ivanmrsulja reviewed (commented):
Looks good to me, but if you have time, please address some of my comments.

litvinovg commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/admin/RolesController.java:
```diff
@@ -0,0 +1,265 @@
+            m.remove(toRemove);
+        }
+        PolicyTemplateController.removeRoleDataSets(uri);
+        // TODO: Remove policies related to removed role
```
Just a leftover. Removed.

litvinovg commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/admin/RolesController.java:
```diff
@@ -0,0 +1,265 @@
+        if (isNotValid(uri)) {
+            TemplateResponseValues response = processList(vreq);
+            response.put("errorMessage", I18n.text(vreq, "edited_role_uri_is_invalid", new Object[] { uri }));
+            log.error(String.format("Uri '%s' of role to edit is not valid", uri));
```
Sure.

litvinovg commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/admin/RolesController.java:
```diff
@@ -0,0 +1,265 @@
+        String uriSuffix = label.toUpperCase().replaceAll(" ", "_").replaceAll("[^a-zA-Z0-9_]+", "");
+        String uri = VitroVocabulary.VITRO_AUTH + "ROLE_" + uriSuffix;
+        Random random = new Random();
+        while (userAccounts.containsResource(new ResourceImpl(uri))) {
```
No. Not suitable for this case.

litvinovg commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/controller/admin/RolesController.java:
```diff
@@ -0,0 +1,265 @@
+        return response;
+    }
+
+    private boolean isNotValid(String uri) {
```
In-lined.

chenejac commented on webapp/src/main/webapp/templates/freemarker/body/roles/userRoles-list.ftl:
```diff
@@ -0,0 +1,42 @@
+<section id="show-auth" role="region">
+	<#assign rolesControllerUrl = urls.base + "/admin/roles" >
+	<script type="text/javascript">
+		function editRole(roleUri, roleLabel){
+		  let newLabel = prompt("${i18n().role_rename_prompt?js_string}", roleLabel);
+		  if (newLabel && !(/^\s*$/.test(newLabel))) {
+		    let form = document.forms["edit-" + roleUri];
+		    form.label.value = newLabel;
+		    form.submit();
+		  }
+		}
```
I don't like the approach for editing role names based on a prompt box. Any chance to make a small form for this? I think we usually used this approach in the VIVO application.

litvinovg replied:
Entry forms work well for data stored in content, not so much for configuration graphs.

chenejac reviewed (changes requested):
@litvinovg a couple of comments related to the user roles page:
1. There is the option for deletion even next to the default 5 roles. However, removing of those roles is not working for me. Only, newly added roles can be deleted in my case.  
2. In the case one role of initial 5 roles have been renamed, if someone click on delete it is not removed, the name has been returned to the initial name. 
3. There is no option for controlling the order of roles. It would be nice to have. Sometimes relation between roles could represent subset/superset of permissions, and in this case the order of roles might help the end user in defining permissions (allowed operations).

litvinovg commented:
> @litvinovg a couple of comments related to the user roles page:
> 
>     1. There is the option for deletion even next to the default 5 roles. However, removing of those roles is not working for me. Only, newly added roles can be deleted in my case.
This is because default roles are loaded into in-memory graph and as a result aren't modifiable. I think it is a positive side effect that those roles can't be removed.
>     2. In the case one role of initial 5 roles have been renamed, if someone click on delete it is not removed, the name has been returned to the initial name.
It happens for the same reason.
>     3. There is no option for controlling the order of roles. It would be nice to have. Sometimes relation between roles could represent subset/superset of permissions, and in this case the order of roles might help the end user in defining permissions (allowed operations).
This PR doesn't introduce anything to control role ordering.

ivanmrsulja reviewed (approved):
Looks good to me.

chenejac reviewed (changes requested):
@litvinovg removing of basic roles work now, but it is not working adding new roles checkbox for all operations. When a user add new role, i.e. librarian, and open edit a class page, the new role librarian is only available for display operation (not for update, publish, etc.).

litvinovg commented:
It is due to regressions in recently merged https://github.com/vivo-project/Vitro/pull/484
I fixed the problem you have noticed by adding type for data set key.
Also fixed errors on pages to create new property or a class.

chenejac reviewed (approved):
Well done
